climada/entity/exposures/litpop/litpop.py

```diff
@@ -0,0 +1,1199 @@
+            iso3alpha (e.g. 'JPN'), iso3num (e.g. 92) or name (e.g. 'Togo')
+        res_arcsec : float (optional)
+            Horizontal resolution in arc-sec.
+            The default is 30 arcsec, this corresponds to roughly 1 km.
+        exponents : tuple of two integers, optional
+            Defining power with which lit (nightlights) and pop (gpw) go into LitPop.
+            Default: (1, 1)
+        fin_mode : str, optional
+            Socio-economic value to be used as an asset base that is disaggregated
+            to the grid points within the country
+            * 'pc': produced capital (Source: World Bank), incl. manufactured or
+                    built assets such as machinery, equipment, and physical structures
+                    (pc is in constant 2014 USD)
+            * 'pop': population count (source: GPW, same as gridded population)
+            * 'gdp': gross-domestic product (Source: World Bank)
+            * 'income_group': gdp multiplied by country's income group+1
+            * 'nfw': non-financial wealth (Source: Credit Suisse, of households only)
+            * 'tw': total wealth (Source: Credit Suisse, of households only)
+            * 'norm': normalized by country
+            * 'none': LitPop per pixel is returned unchanged
+            The default is 'pc'.
+        admin1_calc : boolean, optional
+            If True, distribute admin1-level GDP (if available). Default: False
+        reference_year : int, optional
+            Reference year for data sources. Default: 2020
+        gpw_version : int, optional
+            Version number of GPW population data.
+            The default is GPW_VERSION
+        data_dir : Path, optional
+            redefines path to input data directory. The default is SYSTEM_DIR.
+        reproject_first : boolean, optional
+            First reproject nightlight (Lit) and population (Pop) data to target
+            resolution before combining them as Lit^m * Pop^n?
+            The default is True. Warning: Setting this to False affects the
+            disaggregation results - expert choice only
+
+        Raises
+        ------
+        NotImplementedError
+
+        Returns
+        -------
+        None.
+
+        """
+        shape_list = _shape_to_list(shape)
+        # init countries' exposure:
+        self.set_countries(countries, res_arcsec=res_arcsec, exponents=exponents,
+                           fin_mode=fin_mode, reference_year=reference_year,
+                           gpw_version=gpw_version, data_dir=data_dir,
+                           reproject_first=reproject_first)
+
+        # loop over shapes and cut out exposure GDFs within each shape, combine:
+        for idx, shp in enumerate(shape_list):
+            if idx==0:
+                gdf = self.gdf.loc[self.gdf.geometry.within(shp)]
+            else:
+                gdf.append(self.gdf.loc[self.gdf.geometry.within(shp)],
+                               ignore_index=True)
```

Comment:
I would like to work entirely with MultiPolygons, but problems arise when using `shapefile.Shape` instances. These are important because that's the shape format returned by `u_coord.get_admin1_info` (wrapper around `cartopy.shapereader`) which is used to get admin1-level shapes. `shapefile.Shape` is not recognized by `gdf.geometry.within`. Therefore I need to transform it.

Currently, I am aware of two ways of transofrming these shapes but both cause errors:
(1) create `MultiPolygon` from list of `shapely.Polygon`s from a `Shape`object (variable name `shape`) with:
`MulitPolygon([Polygon(shape.points[shape.parts[i]:part-1]) for i, part in enumerate(list(shape.parts[1:])+[0])])`

(2) transfrom into `Polygon` from points ignoring the `parts`: 
`Polygon(shape.points)`



Option (1) works for a shape with multiple sub-shapes (e.g. Schleswig-Holstein with its islands), but not for shapes with holes (e.g. Brandenburg excluding enclave Berlin) - here the problem you mention arises, Berlin is double counted instead of excluded.

(1a) Schleswig-Holstein with first apoproach (correct):
![grafik](https://user-images.githubusercontent.com/27805268/132337977-263529d3-26dd-4680-b60e-3afc2596ca92.png)

(1b) Brandenburg with first approach (incorrect):
![grafik](https://user-images.githubusercontent.com/27805268/132337698-ef7d255c-4511-422e-9a01-fe2c65b3bc5c.png)


For option (2), it is the other way around, see example for Schlesweig Holstein and Brandenburg with 

(2a) Schleswig-Holstein with `Polygon(admin1_shapes[idx].points)` (incorrect):
![grafik](https://user-images.githubusercontent.com/27805268/132337361-b5199d06-99c5-456f-b348-89bf2104e63b.png)
(2b) Brandenburg with `Polygon(admin1_shapes[idx].points)` (ok-ish):
![grafik](https://user-images.githubusercontent.com/27805268/132337481-2ac04a74-edd2-45b8-a6ba-d465d74f8cf9.png)


Now, what I am lookin for is a clean way to transform `shapefile.Shape` to `shapely.MultiPolygon` in a way that exclaves like islands are included properly and enclaves are excluded. Any ideas?

Reply:
Code example:

```
from shapely.geometry import Polygon
from shapely.geometry import MultiPolygon
import climada.util.coordinates as u_coord

country_iso3a = 'DEU'
state_name = 'Brandenburg'
# state_name = 'Schleswig-Holstein'

"""First, we get the shape of from util.coordinates, which gets it from cartopy's shapereader:"""
admin1_info, admin1_shapes = u_coord.get_admin1_info(country_iso3a)
admin1_info = admin1_info[country_iso3a]
admin1_shapes = admin1_shapes[country_iso3a]
admin1_names = [record['name'] for record in admin1_info]
print(admin1_names)
for idx, name in enumerate(admin1_names):
    if admin1_names[idx]==state_name:
        break
print('Brandenburg index: ' + str(idx))
shape = admin1_names[idx]

poly1 = Polygon(shape.points)
poly2 = MulitPolygon([Polygon(shape.points[shape.parts[i]:part-1]) for i, part in enumerate(list(shape.parts[1:])+[0])])

poly2
```

![grafik](https://user-images.githubusercontent.com/27805268/132338877-f9569f55-c217-4676-b9d8-b97f1156dd63.png)

Reply:
It would be nice to avoid `shapereader.Shape` completely, but there seems to be no field `geometries` in the admin_1_states_provinces data from Natural Earth Data.

Reply:
**Update:**
I found a solution, using a workaround for `set_custom_shape_from_countries`:

```
        self.set_countries(countries, res_arcsec=res_arcsec, exponents=exponents,
                           fin_mode=fin_mode, reference_year=reference_year,
                           gpw_version=gpw_version, data_dir=data_dir,
                           reproject_first=reproject_first)

        if isinstance(shape, list):
        # loop over shapes and cut out exposure GDFs within each shape, combine:
            for idx, shp in enumerate(shape):
                if idx==0:
                    gdf = self.gdf.loc[self.gdf.geometry.within(shp)]
                else:
                    gdf.append(self.gdf.loc[self.gdf.geometry.within(shp)],
                                   ignore_index=True)
        elif isinstance(shape, Shape):
            # get gdf with geometries of points within shape:
            shape_gdf, _ = _get_litpop_single_polygon(shape, reference_year,
                                                      res_arcsec, data_dir,
                                                      gpw_version, reproject_first,
                                                      exponents,
                                                      )
            shape_gdf = shape_gdf.drop(
                columns=shape_gdf.columns[shape_gdf.columns != 'geometry'])
            # extract gdf with data points within shape:
            gdf = geopandas.sjoin(self.gdf, shape_gdf, how='right')
        else: # works if shape is Polygon or MultiPolygon
            gdf = self.gdf.loc[self.gdf.geometry.within(shape)]
```

With this work around, and avoiding transforming the shape into a list for `set_custom_shape`, the problem is solved.

Reply:
c.f. commit : https://github.com/CLIMADA-project/climada_python/pull/267/commits/14088d4df98d748d0bf9da637ec2b1297c4ab1cd